File: cpp/1408.string-matching-in-an-array.cpp
#include <algorithm>
#include <string>
#include <vector>

using std::string;
using std::vector;

const int ALPHA = 26;
class TrieNode {
public:
  TrieNode *child[ALPHA];
};

class Trie {
private:
  TrieNode *_root;

public:
  Trie() : _root(nullptr) {}

  void addWord(const string &word) {
    if (!_root) {
      _root = new TrieNode();
    }

    auto node = _root;
    for (char ch : word) {
      if (!node->child[ch - 'a'])
        node->child[ch - 'a'] = new TrieNode();
      node = node->child[ch - 'a'];
    }
  }

  bool isSubstr(const string &word) {
    if (!_root) {
      return false;
    }

    auto node = _root;
    for (char ch : word) {
      if (!node->child[ch - 'a'])
        return false;
      node = node->child[ch - 'a'];
    }

    return true;
  }
};

class Solution {
public:
  vector<string> stringMatching(vector<string> &words) {

    std::sort(words.begin(), words.end(),
              [](string &a, string &b) -> bool { return a.size() > b.size(); });

    vector<string> res;
    Trie trie;

    for (string &word : words) {
      if (trie.isSubstr(word)) {
        res.push_back(word);
      }

      // add all possible substrings
      for (int i = 0; i < word.size(); i++) {
        trie.addWord(word.substr(i));
      }
    }

    return res;
  }
};

File: golang/1408.string-matching-in-an-array.go

func stringMatching(words []string) []string {
	trie := NewTrie()

	res := make([]string, 0)

	for _, word := range words {
		for i := 0; i < len(word); i++ {
			trie.AddWord(word[i:])
		}
	}

	for _, word := range words {
		if trie.IsSubstr(word) {
			res = append(res, word)
		}
	}

	return res
}

type TrieNode struct {
	Child [26]*TrieNode
	Count int
}

type Trie struct {
	root *TrieNode
}

func NewTrie() Trie {
	return Trie{root: &TrieNode{}}
}

func (t *Trie) AddWord(word string) {
	node := t.root

	for i := 0; len(word) > i; i++ {
		if node.Child[word[i]-'a'] == nil {
			node.Child[word[i]-'a'] = &TrieNode{}
		}

		node = node.Child[word[i]-'a']
		node.Count++
	}
}

func (t *Trie) IsSubstr(word string) bool {
	node := t.root

	for i := 0; len(word) > i; i++ {
		if node.Child[word[i]-'a'] == nil {
			return false
		}
		node = node.Child[word[i]-'a']
	}

	return node.Count > 1
}
